### PR TITLE
Probability Distribution and Statistical Functions -- Normal Distribution Module 

### PR DESCRIPTION
**Pull request by Jim-215-Fisher**
This is the second round of probability distribution and statistical functions, continuing from #240. Since the whole modular structure has been changed, each distribution will have its own PR.
This PR contains normal distributions. Files uploaded are:
src/stdlib_stats_distribtuion_normal.fypp
src/CmakeLists.txt
src/Makefile.manual
src/tests/test_distribution_normal.f90
src/tests/CmakeLists.txt
src/tests/Makefile.manual
doc/specs/stdlib_stats_distribution_normal.md
doc/specs/index.md

### CODE REVIEW

**Comment by ivan-pi** [comment]
Hello @Jim-215-Fisher ,

it's an impressive undertaking what you have here!

I just gave the files a quick glance and noticed certain calls to `real(val)` might not be what was desired. To preserve the precision of the real part of a complex number you need to use the full syntax: `real(cval,kind= ...)`. An alternative is to use `cval%re`. A discussion of this pitfall can be found [here](https://fortran-lang.discourse.group/t/the-counter-intuitive-rise-of-python-in-scientific-computing/469/38?u=ivanpribec).

**Comment by Jim-215-Fisher** [comment]
> Hello @Jim-215-Fisher ,
> 
> it's an impressive undertaking what you have here!
> 
> I just gave the files a quick glance and noticed certain calls to `real(val)` might not be what was desired. To preserve the precision of the real part of a complex number you need to use the full syntax: `real(cval,kind= ...)`. An alternative is to use `cval%re`. A discussion of this pitfall can be found [here](https://fortran-lang.discourse.group/t/the-counter-intuitive-rise-of-python-in-scientific-computing/469/38?u=ivanpribec).

Thanks for the point. I will take a look of it. Originally, I was using cval%re and cval%im, but it did not pass the compilation on gfortran 7,8,9.

**Comment by Jim-215-Fisher** [comment]
> Hello @Jim-215-Fisher ,
> 
> it's an impressive undertaking what you have here!
> 
> I just gave the files a quick glance and noticed certain calls to `real(val)` might not be what was desired. To preserve the precision of the real part of a complex number you need to use the full syntax: `real(cval,kind= ...)`. An alternative is to use `cval%re`. A discussion of this pitfall can be found [here](https://fortran-lang.discourse.group/t/the-counter-intuitive-rise-of-python-in-scientific-computing/469/38?u=ivanpribec).

I checked the Fortran 2018 Handbook, it states like this for real(A, [KIND]):
"if A is type complex and KIND is not present, the kind type parameter is the kind type parameter of A."
This is the result I desired in the code. Maybe I missed something.

**Comment by ivan-pi** [comment]
> I checked the Fortran 2018 Handbook, it states like this for real(A, [KIND]):
> "if A is type complex and KIND is not present, the kind type parameter is the kind type parameter of A."
> This is the result I desired in the code. Maybe I missed something.

In that case consider my comment void. I was basing my comment of the comments I read on Fortran Discourse.

**Comment by jvdp1** [comment]
Wrong button....I re-open this PR.

**Comment by jvdp1** [comment]
Thank you @Jim-215-Fisher for the changes. I don't have additional major changes currently.
@ivan-pi @awvwgk @milancurcic @gareth-nx  could some of you review this PR to try to move it foward please?

File: src/tests/stats/test_distribution_normal.fypp
Comment (awvwgk) [inline src/tests/stats/test_distribution_normal.fypp:5]: Can you make this import explicit?

File: src/tests/stats/test_distribution_normal.fypp
Comment (awvwgk) [inline src/tests/stats/test_distribution_normal.fypp:6]: Might be worth to switch to using test-drive's `check` function for tests

**Review by awvwgk — APPROVED** [review]
Would be nice to switch the tests to the new framework before merging, but this can also happen in a separate PR.

Looks good to me.

File: doc/specs/stdlib_stats_distribution_normal.md
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_normal.md:107]: Does this definition also cover the case of complex `x`? I am not immediately familiar with the normal density for a complex random variable, so it would be good to provide some reference.

File: doc/specs/stdlib_stats_distribution_normal.md
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_normal.md:190]: Does this directly cover the complex case? In any case it would be good to provide a reference that covers that.

File: src/tests/stats/test_distribution_normal.fypp
Comment (Jim-215-Fisher) [reply src/tests/stats/test_distribution_normal.fypp:5]: OK

File: src/tests/stats/test_distribution_normal.fypp
Comment (Jim-215-Fisher) [reply src/tests/stats/test_distribution_normal.fypp:6]: Don't know what is test-drive. Where is it located?

File: doc/specs/stdlib_stats_distribution_normal.md
Comment (Jim-215-Fisher) [reply doc/specs/stdlib_stats_distribution_normal.md:107]: I will elaborate the formula for the complex.

File: src/tests/stats/test_distribution_normal.fypp
Comment (jvdp1) [reply src/tests/stats/test_distribution_normal.fypp:6]: See  e.g. [this example](https://github.com/fortran-lang/stdlib/blob/master/src/tests/stats/test_mean.fypp).
However, since the results of the different functions depend on the state of different global variables, it might not work until [this PR](https://github.com/fortran-lang/test-drive/issues/10) (or another similar) is merged.

for the moment, I propose to keep them as they are and to change them in another PR.

**Comment by Jim-215-Fisher** [comment]
@jvdp1 @awvwgk @gareth-nx , the files have been updated. Especially documentation for complex variable case was changed.

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [inline src/stdlib_stats_distribution_normal.fypp:277]: Here I think `res` should be of type ${t1}$ ? I noticed differences between this function and the results from Rs `dnorm`, when usign double-precision in Fortran, that are likely due to this.

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [inline src/stdlib_stats_distribution_normal.fypp:293]: Again, should `res` be of type `${t1}$` ?

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [inline src/stdlib_stats_distribution_normal.fypp:309]: specify type of `res`.

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [inline src/stdlib_stats_distribution_normal.fypp:324]: as above, type of `res`

Comment (gareth-nx) [reply src/stdlib_stats_distribution_normal.fypp:324]: Actually in this case, I believe res should be real (but with the same kind as the real part of `${t1}$`

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [reply src/stdlib_stats_distribution_normal.fypp:293]: No, I think `res` should be real but with the same kind as `${t1}$` (??).

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [reply src/stdlib_stats_distribution_normal.fypp:277]: Note: The same kind of bug also occurs in the equivalent `uniform` module

Comment (gareth-nx) [reply src/stdlib_stats_distribution_normal.fypp:277]: Just confirming that when I make the obvious fix in my local branch of stdlib, the results agree with R to within `5e-016`, when checking 401 evenly spaced points from -10 to 10 standard deviations in double precision. Whereas the differences reach around `1e-08` otherwise.

Comment (gareth-nx) [reply src/stdlib_stats_distribution_normal.fypp:277]: It would be good if the unit tests were able to catch such precision issues (easy mistake to make or introduce in future when editing).

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [inline src/stdlib_stats_distribution_normal.fypp:314]: While correct, this representation is prone to round-off error when the `erf` term is close to -1.0 - which in practice leads to greater relative errors when `x` is quite a few standard deviations below the mean.

To improve the results in that range of x, consider instead using the intrinsic `erfc`. I think the equation is `0.5*erfc(-(x-loc)/(scale*sqrt_2))`.

File: src/stdlib_stats_distribution_normal.fypp
Comment (Jim-215-Fisher) [reply src/stdlib_stats_distribution_normal.fypp:277]: May be due to my limited probability knowledge, I purposely restrict the results of all probability function value to be single precision. To me, the **probability** of 0.1234567 (single precision) would be the same as **probability** of 0.12345678901234567 (double precision) practically, thought mathematically there are different. May be I am wrong.

Comment (Jim-215-Fisher) [reply src/stdlib_stats_distribution_normal.fypp:277]: Should we provide all precision for probability?

Comment (gareth-nx) [reply src/stdlib_stats_distribution_normal.fypp:277]: Yes, we should provide all precision for all probability. For these kids of functions, the need for precision is similar to other branches of applied math. 

While in some cases single-precision is fine, we also do things that need higher precision (e.g. suppose these functions appear in an expression that is then numerically differentiated).

I think the solution is to make `res` be of kind `real(${k1}$)` in all of the functions. Casual tests of `pdf_uniform` and `cdf_uniform` with this approach gave results very close to R's `dnorm` and `pnorm` in double precision.

File: src/tests/stats/test_distribution_normal.fypp
Comment (gareth-nx) [inline src/tests/stats/test_distribution_normal.fypp:156]: I think here, if the numbers are updated with quad-precision results, then the tests should be able to catch errors caused by loss of precision. They have a precision-dependent tolerance already - it might need tweaking if extended to higher precisions, but looks like it should just work.

File: src/tests/stats/test_distribution_normal.fypp
Comment (gareth-nx) [inline src/tests/stats/test_distribution_normal.fypp:163]: As above, update RHS with quad-precision results, and use `real(${k1}$), parameter :: ans`

File: src/tests/stats/test_distribution_normal.fypp
Comment (gareth-nx) [inline src/tests/stats/test_distribution_normal.fypp:203]: As above, this can be modified to catch loss of precision in the results

File: src/tests/stats/test_distribution_normal.fypp
Comment (gareth-nx) [inline src/tests/stats/test_distribution_normal.fypp:211]: As above

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [reply src/stdlib_stats_distribution_normal.fypp:293]: Based on tests I ran just now, it should be: `real(${k1}$) :: res`

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [reply src/stdlib_stats_distribution_normal.fypp:277]: Based on tests I ran just now, it should be: `real(${k1}$) :: res`

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [reply src/stdlib_stats_distribution_normal.fypp:309]: Based on tests I ran just now, it should be: `real(${k1}$) :: res`

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [reply src/stdlib_stats_distribution_normal.fypp:314]: Just now I tried changing this to be 
```
        !res = (1.0_${k1}$ + erf((x - loc) / (scale * sqrt_2))) / 2.0_${k1}$
        res = erfc(-(x - loc) / (scale * sqrt_2)) / 2.0_${k1}$
```
All the tests passed, and a comparison with R showed much better agreement when `x` was around -8 standard deviations below `loc`, where we expect the previous formulation to have trouble.

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [inline src/stdlib_stats_distribution_normal.fypp:282]: Here, to reduce the risk of numerical problems when `x, loc, scale` are all very large magnitude, perhaps it would be better to write this as `exp(- 0.5_${k1}$ * ((x - loc)/scale) * ((x - loc) /scale))) `

File: src/stdlib_stats_distribution_normal.fypp
Comment (Jim-215-Fisher) [reply src/stdlib_stats_distribution_normal.fypp:277]: So here we got a problem. The normal distribution is based on Marsaglia's fast ziggurat method which use double precision for the ziggurat table. It would be OK for the single and double precision probability but quardruple precision probability. For quardruple precision probability, any number beyond 16 significant figures will be meaningless.

Comment (gareth-nx) [reply src/stdlib_stats_distribution_normal.fypp:277]: I see, thanks. In the first instance, I think it's OK if the random number generator does not give quad-precision accuracy - we can worry about that later if it's important. But for the pdf/cdf functions, we should be able to get full precision without any problems.

Comment (gareth-nx) [reply src/stdlib_stats_distribution_normal.fypp:277]: But please add a statement in the specs about the precision of the random numbers being fundamentally limited to double precision by the algorithm (even if results are provided for quad precision).

File: src/stdlib_stats_distribution_normal.fypp
Comment (Jim-215-Fisher) [reply src/stdlib_stats_distribution_normal.fypp:314]: Thanks.

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [inline src/stdlib_stats_distribution_normal.fypp:12]: Here I suggest you add a comment `! These variables are only used in the random number generation, which is fundamentally double precision`.

File: doc/specs/stdlib_stats_distribution_normal.md
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_normal.md:222]: Can be updated now that more precisions are supported.

File: doc/specs/stdlib_stats_distribution_normal.md
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_normal.md:135]: Can be updated now that more precisions are supported

File: doc/specs/stdlib_stats_distribution_normal.md
Comment (gareth-nx) [reply doc/specs/stdlib_stats_distribution_normal.md:47]: I suggest also mentioning that the internal computations produce double precision random numbers, so higher precision outputs are simply type-cast from these and will not have full-precision accuracy.

File: doc/specs/stdlib_stats_distribution_normal.md
Comment (gareth-nx) [reply doc/specs/stdlib_stats_distribution_normal.md:19]: I believe more precisions are supported now, and this can be updated (?).

**Comment by Jim-215-Fisher** [comment]
@jvdp1 @awvwgk @gareth-nx All precision for probability were added according to @gareth-nx . Once this PR is done, I will set up a new PR for uniform distribution to correct similar problems.

**Review by gareth-nx — COMMENTED** [review]
Thankyou! This is a very important contribution to stdlib.

File: src/stdlib_stats_distribution_normal.fypp
Comment (gareth-nx) [reply src/stdlib_stats_distribution_normal.fypp:282]: @Jim-215-Fisher what do you think about this change?

Comment (Jim-215-Fisher) [reply src/stdlib_stats_distribution_normal.fypp:282]: Sorry, did not see it.

**Comment by jvdp1** [comment]
@Jim-215-Fisher @gareth-nx  is this PR ready for merging?

**Comment by gareth-nx** [comment]
@jvdp1 @Jim-215-Fisher I think it just needs some minor documentation fixes mentioned above (or, a comment about why they are not needed -- I haven't looked in detail since reviewing). I was intending to approve once those points were addressed.

**Comment by Jim-215-Fisher** [comment]
> @Jim-215-Fisher @gareth-nx is this PR ready for merging?

I think it is ready.

**Comment by Jim-215-Fisher** [comment]
> @jvdp1 @Jim-215-Fisher I think it just needs some minor documentation fixes mentioned above (or, a comment about why they are not needed -- I haven't looked in detail since reviewing). I was intending to approve once those points were addressed.

I don't think we mention precision in documentation though. By default, all precision are supported.

**Comment by gareth-nx** [comment]
> > @jvdp1 @Jim-215-Fisher I think it just needs some minor documentation fixes mentioned above (or, a comment about why they are not needed -- I haven't looked in detail since reviewing). I was intending to approve once those points were addressed.
> 
> I don't think we mention precision in documentation though. By default, all precision are supported.

Ahh, I see. I am getting confused between type and kind in Fortran. 

So at the moment the documentation indicates the output type, but not the kind, and it is assumed to be implied that all kinds are supported.

I think it would be nicer to be more explicit about this. The same point was made in [this review comment](https://github.com/fortran-lang/stdlib/pull/273/files/e786bd501fb6cdfac80fc48512081550eb93ae85#r725516236) by @jvdp1 on October 10. @Jim-215-Fisher would you mind to use phrasing like this when describing the output variables?

**Comment by Jim-215-Fisher** [comment]
> > > @jvdp1 @Jim-215-Fisher I think it just needs some minor documentation fixes mentioned above (or, a comment about why they are not needed -- I haven't looked in detail since reviewing). I was intending to approve once those points were addressed.
> > 
> > 
> > I don't think we mention precision in documentation though. By default, all precision are supported.
> 
> Ahh, I see. I am getting confused between type and kind in Fortran.
> 
> So at the moment the documentation indicates the output type, but not the kind, and it is assumed to be implied that all kinds are supported.
> 
> I think it would be nicer to be more explicit about this. The same point was made in [this review comment](https://github.com/fortran-lang/stdlib/pull/273/files/e786bd501fb6cdfac80fc48512081550eb93ae85#r725516236) by @jvdp1 on October 10. @Jim-215-Fisher would you mind to use phrasing like this when describing the output variables?

done.

**Review by gareth-nx — APPROVED** [review]
Many thanks!

**Comment by gareth-nx** [comment]
@jvdp1 This is ready to merge from my view.

**Review by jvdp1 — APPROVED** [review]
LGTM. Thank you @Jim-215-Fisher for this and to all reviewers. i'll merge it.

**Pull request merged**
